src/Database/Statement.php

```diff
@@ -0,0 +1,370 @@
+     *
+     * @param array $params list of values to be bound
+     * @param array $types list of types to be used, keys should match those in $params
+     * @return void
+     */
+    public function bind(array $params, array $types): void
+    {
+        if (empty($params)) {
+            return;
+        }
+
+        $anonymousParams = is_int(key($params));
+        $offset = 1;
+        foreach ($params as $index => $value) {
+            $type = $types[$index] ?? null;
+            if ($anonymousParams) {
+                /** @psalm-suppress InvalidOperand */
+                $index += $offset;
+            }
+            /** @psalm-suppress PossiblyInvalidArgument */
+            $this->bindValue($index, $value, $type);
+        }
+    }
+
+    /**
+     * Assign a value to a positional or named variable in prepared query. If using
+     * positional variables you need to start with index one, if using named params then
+     * just use the name in any order.
+     *
+     * It is not allowed to combine positional and named variables in the same statement.
+     *
+     * ### Examples:
+     *
+     * ```
+     * $statement->bindValue(1, 'a title');
+     * $statement->bindValue('active', true, 'boolean');
+     * $statement->bindValue(5, new \DateTime(), 'date');
+     * ```
+     *
+     * @param string|int $column name or param position to be bound
+     * @param mixed $value The value to bind to variable in query
+     * @param string|int|null $type name of configured Type class
+     * @return void
+     */
+    public function bindValue(string|int $column, mixed $value, string|int|null $type = 'string'): void
+    {
+        if ($type === null) {
+            $type = 'string';
+        }
+        if (!is_int($type)) {
+            [$value, $type] = $this->cast($value, $type);
+        }
+
+        $this->params[$column] = $value;
+        $this->performBind($column, $value, $type);
+    }
+
+    /**
+     * Perform bind to statement.
+     *
+     * @param string|int $column name or param position to be bound
+     * @param mixed $value The value to bind to variable in query
+     * @param int $type PDO::PARAM_* constant
+     * @return void
+     */
+    protected function performBind(string|int $column, mixed $value, int $type): void
+    {
+        $this->statement->bindValue($column, $value, $type);
+    }
+
+    /**
+     * Executes the statement by sending the SQL query to the database. It can optionally
+     * take an array or arguments to be bound to the query variables. Please note
+     * that binding parameters from this method will not perform any custom type conversion
+     * as it would normally happen when calling `bindValue`.
+     *
+     * @param array|null $params list of values to be bound to query
+     * @return bool true on success, false otherwise
+     */
+    public function execute(?array $params = null): bool
+    {
+        $loggedQuery = new LoggedQuery();
```

Comment:
Shouldn't the `LoggedQuery` instance be created only if logging is enabled?

Reply:
I would like to clean up this function, but I didn't figure anything out yet. I do want to avoid a bunch of if checks and temp variables. The requirement to capture and log the error makes things a little clunky.

Reply:
Yes, the currently code is quite clunky.

Reply:
Could we take a similar approach to how we do error logging where the metadata (duration, row counts) is sent to a logger that then decides on whether or not to log the query? That might be a breaking change but again I suspect there are very few database logging implementations in the wild so the impact would be low.

Reply:
Not sure if that would be the right approach. Logging is an optional feature. So if it's not enabled then a logger instance wouldn't/shouldn't exist.

Reply:
Cleaned up logging in execute(). Only runs if logger is set.